packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -476,143 +732,161 @@ class _CupertinoPersistentNavigationBar extends StatelessWidget implements Prefe
+  const _BackChevron({
+    @required this.color,
+  }) : assert(color != null);
+
+  final Color color;
 
   @override
-  bool shouldRebuild(_CupertinoLargeTitleNavigationBarSliverDelegate oldDelegate) {
-    return persistentHeight != oldDelegate.persistentHeight
-        || title != oldDelegate.title
-        || leading != oldDelegate.leading
-        || middle != oldDelegate.middle
-        || trailing != oldDelegate.trailing
-        || border != oldDelegate.border
-        || backgroundColor != oldDelegate.backgroundColor
-        || actionsForegroundColor != oldDelegate.actionsForegroundColor;
+  Widget build(BuildContext context) {
+    final TextDirection textDirection = Directionality.of(context);
+
+    // Replicate the Icon logic here to get a tightly sized icon and add
+    // custom non-square padding.
+    Widget iconWidget = new Text.rich(
+      new TextSpan(
+        text: new String.fromCharCode(CupertinoIcons.back.codePoint),
+        style: new TextStyle(
+          inherit: false,
+          color: color,
+          fontSize: 34.0,
+          fontFamily: CupertinoIcons.back.fontFamily,
+          package: CupertinoIcons.back.fontPackage,
+        ),
+      ),
+    );
+    switch (textDirection) {
+      case TextDirection.rtl:
+        iconWidget = new Transform(
+          transform: new Matrix4.identity()..scale(-1.0, 1.0, 1.0),
+          alignment: Alignment.center,
+          transformHitTests: false,
+          child: iconWidget,
+        );
+        break;
+      case TextDirection.ltr:
+        break;
+    }
+
+    return iconWidget;
+  }
+}
+
+/// A widget that shows next to the back chevron when `automaticallyImplyLeading`
+/// is true.
+class _BackLabel extends StatelessWidget {
+  const _BackLabel({
+    @required this.specifiedPreviousTitle,
+    @required this.route,
+  }) : assert(route != null);
+
+  final String specifiedPreviousTitle;
+  final ModalRoute<dynamic> route;
+
+  // `child` is never passed in into ValueListenableBuilder so it's always
+  // null here and unused.
+  Widget _buildPreviousTitleWidget(BuildContext context, String previousTitle, Widget child) {
+    if (previousTitle?.isNotEmpty == true) {
+      if (previousTitle.length > 10) {
+        return const Text('Back');
+      } else {
+        return new Text(previousTitle, maxLines: 1);
+      }
+    } else {
+      return const SizedBox(height: 0.0, width: 0.0);
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (specifiedPreviousTitle?.isNotEmpty == true) {
+      return _buildPreviousTitleWidget(context, specifiedPreviousTitle, null);
+    } else if (route is CupertinoPageRoute<dynamic>) {
+      final CupertinoPageRoute<dynamic> cupertinoRoute = route;
+      return new ValueListenableBuilder<String>(
+        valueListenable: cupertinoRoute.previousTitle,
+        builder: _buildPreviousTitleWidget,
+      );
+    } else {
+      return const SizedBox(height: 0.0, width: 0.0);
+    }
```

Comment:
lots of returns-after-else in the last 30 lines

Reply:
Done. I left the if-else if-else at the end which seems more compact than a series of exits